blockchain/state_manager/src/lib.rs

```diff
@@ -223,10 +262,238 @@ where
+                    first_subscriber
+                )))
+            }
+        };
+        let maybe_message_reciept = self.tipset_executed_message(&tipset, cid, &*message)?;
+        if let Some(r) = maybe_message_reciept {
+            if let ChainMessage::HcCurrent(current) = first_subscriber {
+                return Ok(Some((current.clone(), r)));
+            }
+        }
+
+        let (back_tipset, _back_receipt) = self
+            .search_back_for_message(&tipset, &*message)?
+            .ok_or_else(|| {
+                Error::Other("State manager not subscribed to back search wait".to_string())
+            })?;
+
+        let mut candidate_tipset: Option<Arc<Tipset>> = None;
+        let mut candidate_receipt: Option<MessageReceipt> = None;
+        let height_of_head = tipset.epoch();
+        let mut reverts: HashMap<TipsetKeys, bool> = HashMap::new();
+        loop {
+            while let Some(subscriber) = self
+                .subscriber
+                .clone()
+                .ok_or_else(|| {
+                    Error::Other("State Manager not subscribed to tipset head changes".to_string())
+                })?
+                .next()
+                .await
+            {
+                match subscriber {
+                    ChainMessage::HcRevert(_tipset) => {
+                        if candidate_tipset.is_some() {
+                            candidate_tipset = None;
+                            candidate_receipt = None;
+                        }
+                    }
+                    ChainMessage::HcApply(tipset) => {
+                        if candidate_tipset
+                            .as_ref()
+                            .map(|s| s.epoch() >= s.epoch() + tipset.epoch())
+                            .unwrap_or_default()
+                        {
+                            let ts = candidate_tipset
+                                .ok_or_else(|| Error::Other("Candidate Tipset not".to_string()))?;
+
+                            let rs = candidate_receipt.ok_or_else(|| {
+                                Error::Other("Candidate Receipt not set".to_string())
+                            })?;
+
+                            return Ok(Some((ts, rs)));
+                        }
+
+                        reverts.insert(tipset.key().to_owned(), true);
+
+                        let maybe_receipt =
+                            self.tipset_executed_message(&tipset, cid, &*message)?;
+                        if let Some(receipt) = maybe_receipt {
+                            if confidence == 0 {
+                                return Ok(Some((tipset, receipt)));
+                            }
+                            candidate_tipset = Some(tipset);
+                            candidate_receipt = Some(receipt)
+                        }
+                    }
+                    _ => continue,
+                }
+            }
+
+            match back_search_wait.next().await {
```

Comment:
If i understand correctly, this is waiting for the back search to complete essentially right? I dont understand why you need to have a subscriber as part of the struct. IT seems like when you start the search_back_for_msg you can create a one shot channel there and run the search back in a blocking task and it will send a notification through the one shot channel when it is done.

Reply:
It looked like almost the search_back_for_msg was being sent in externally? isMostly because all events are coming from tsub which is subheadchanges? What do you think of this?

Reply:
Ok, i think i understand more whats happening